### PR TITLE
feat(backup): aws backup for s3 buckets

### PR DESCRIPTION
**Pull request by PeterBaker0**
This is my first contribution to the wonderful CDK - so apologies if I've missed something. 

This PR is a non breaking change which adds support for S3 buckets to the AWS Backup L2 constructs. 

This was realised by

- a `fromS3Bucket` static method on the `BackupResource` construct
- a check for the bucket type in the Backupable Resources Collector Aspect

When using this construct for a project, I found two other issues 

1. The `BackupSelection` construct does not add the S3 backup/restore managed policies and has no means to do so
2. The `BackupSelection` construct does not expose the IAM role (at least directly - it was previously set as the `grantPrincipal`)

So this PR also:

1. Adds `allowS3Backup` and `allowS3Restores` flags which add the AWS Backup for S3 managed policies for backup/restore respectively 
2. Exposes the `role` in the BackupSelection construct

Additionally, I found that in using this construct, it was not possible to provide a fully customised role to the construct, it would always add, at least, the managed policy for backup. So I added a `customRole` flag which can be provided to disable the addition of any managed policies, meaning that a user could provide a completely custom role, if desired. Annotation warnings for the construct were added in the case where both the `customRole` flag and any of the `allowX` flags for managed policies were both `true`.

An integration test was added which makes sure that selections both using the `fromS3Bucket` and `fromConstruct` resource static methods function. The `allowS3Backup` and `allowS3Restore` flags were true to ensure that the managed policy names were correct. 

Added unit tests for the S3 support (static resource methods, managed policies), the IAM role property, and the `customRole` 
flag behaviour. 

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-backup/README.md
Comment (Naumel) [inline packages/@aws-cdk/aws-backup/README.md:86]: `Fake` sounds harsh, why not `Managed`?

File: packages/@aws-cdk/aws-backup/README.md
Comment (Naumel) [inline packages/@aws-cdk/aws-backup/README.md:201]: 😄 good catch

File: packages/@aws-cdk/aws-backup/README.md
Comment (Naumel) [inline packages/@aws-cdk/aws-backup/README.md:242]: Why `user` in the sentence? Can it not be skipped completely?

File: packages/@aws-cdk/aws-backup/README.md
Comment (Naumel) [inline packages/@aws-cdk/aws-backup/README.md:245]: An extra reference never hurts, e.g. https://docs.aws.amazon.com/aws-backup/latest/devguide/iam-service-roles.html

File: packages/@aws-cdk/aws-backup/README.md
Comment (Naumel) [inline packages/@aws-cdk/aws-backup/README.md:255]: 👌

File: packages/@aws-cdk/aws-backup/README.md
Comment (Naumel) [inline packages/@aws-cdk/aws-backup/README.md:262]: This is just me being extra, but if you find it in your heart to add the code samples going from one option to the other, that's be 🙌

File: packages/@aws-cdk/aws-backup/lib/selection.ts
Comment (Naumel) [inline packages/@aws-cdk/aws-backup/lib/selection.ts:147]: Should the customer role not be set somewhere though, as per
>So I added a customRole flag which can be provided to disable the addition of any managed policies, meaning that a user could provide a completely custom role, if desired.

**Review by Naumel — DISMISSED** [review]
Thank you for taking the time and submitting a PR.
Most things look good and this is a nice addition.

I have a number for positive remarks and questions scattered through your change. 

The most pressing one I will write here, please add unit tests, especially since there are misconfiguration cases.

👍 thank you for the description too, really makes contextualization easy.

File: packages/@aws-cdk/aws-backup/README.md
Comment (PeterBaker0) [reply packages/@aws-cdk/aws-backup/README.md:86]: Sure :+1:

File: packages/@aws-cdk/aws-backup/README.md
Comment (PeterBaker0) [reply packages/@aws-cdk/aws-backup/README.md:242]: Good catch

File: packages/@aws-cdk/aws-backup/README.md
Comment (PeterBaker0) [reply packages/@aws-cdk/aws-backup/README.md:245]: Okay - will add in - thanks

File: packages/@aws-cdk/aws-backup/README.md
Comment (PeterBaker0) [reply packages/@aws-cdk/aws-backup/README.md:262]: Happy to do that - not extra at all :rofl:

File: packages/@aws-cdk/aws-backup/lib/selection.ts
Comment (PeterBaker0) [reply packages/@aws-cdk/aws-backup/lib/selection.ts:147]: Not sure what you mean here - I think maybe `customRole` is not a good name for the field. 

Personally I think the default behaviour of the service role policy being added even when the user provides their own Role is debatable/opinionated - this flag is to add the capacity to have the construct just reference your provided role or create one on your behalf without adding any policies to it. 

Happy for ideas on how to clarify this - but as I said I don't fully understand your comment above.

Comment (PeterBaker0) [reply packages/@aws-cdk/aws-backup/lib/selection.ts:147]: Maybe another approach would be to add an `allowBackup` flag (similar to `allowRestore`) which is default to `true`. Then all of the managed policies can just be switched on/off, but a reasonable default selection of policies is present. This also means there would be no breaking change to the API.

**Comment by PeterBaker0** [comment]
> The most pressing one I will write here, please add unit tests, especially since there are misconfiguration cases.

I did add a few unit tests in the selections test - this covered the `fromS3Bucket`, `fromConstruct`, as well as the changes to the roles. Wasn't sure how to handle the unit test case for misconfiguration - how do you enforce that there should be annotation warnings?

**Comment by PeterBaker0** [comment]
@Naumel thanks for your review. 

I think the `customRole` interface was confusing so changed it to include a `allowBackup` flag which defaults to `true`. This means that there is no breaking change, however users can easily specify `allowBackup` as `false` if they need to use a custom role which doesn't include this managed policy. 

As mentioned above - I did add a few unit tests and update relevant existing ones - could you clarify which features are missing from the unit test coverage if you would like more? Thanks. 

I also added another new integration test to verify that the various combinations of policies are all valid and deploy properly. 

Readme was updated and streamlined. Added a table to map the allowX fields to the managed policies, and supplied a link to relevant section in AWS docs. 

Cheers.

**Review by Naumel — APPROVED** [review]
I like the renaming, makes reading easier for sure.

If you decide to add even more functionality, I look forward to the submission(s)!

**Comment by Naumel** [comment]
@Mergifyio update

**Comment by PeterBaker0** [comment]
> I like the renaming, makes reading easier for sure.
> 
> If you decide to add even more functionality, I look forward to the submission(s)!

Thanks! 

There was a stale reference in the parameter comments so have corrected that now - apologies if that breaks your PR approval!

**Review by Naumel — DISMISSED** [review]

**Comment by PeterBaker0** [comment]
@Naumel I have read the contribution guide but couldn't find any advice on the next steps. Do I need to take any more actions to get this merged or will happen automatically? Cheers.

File: packages/@aws-cdk/aws-backup/README.md
Comment (kaizencc) [inline packages/@aws-cdk/aws-backup/README.md:75]:
```suggestion
`BackupSelection`, see the [Backup selection IAM role policies](#backup-selection-iam-role-policies) section.
```

File: packages/@aws-cdk/aws-backup/README.md
Comment (kaizencc) [inline packages/@aws-cdk/aws-backup/README.md:40]:
```suggestion
const myBucket = new s3.Bucket(this, 'Bucket');
```

File: packages/@aws-cdk/aws-backup/README.md
Comment (kaizencc) [inline packages/@aws-cdk/aws-backup/README.md:278]: Question: would someone ever send in a s3 bucket in `resources` and not want to set `allowS3Backup`? To me, it seems like they go together and if so, we should find some way to determine if a resource is an s3 bucket and automatically add the s3 specific policy.

Same goes for `allowS3Restores`, I feel like we should be able to bake it into `allowRestores` and provide a better user experience.

**Review by kaizencc — COMMENTED** [review]
Hi @PeterBaker0! To answer your question, normally once you have approval you're good to go and a maintainer will figure out how to wrestle with mergify to get your PR in. However, I chanced on this PR after reading the corresponding issues, and I have a few questions reading through this PR. Hope you can indulge me.

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio unqueue

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Putting this back into changes requested given @kaizencc's comments.

**Comment by TheRealAmazonKendra** [comment]
One quick question from me: is there an issue for this feature? If so, please link it so it will get closed when this is merged. If not, no worries.

**Comment by kaizencc** [comment]
@PeterBaker0 are you still working on this PR? I see some 👍 on my previous comments but no action besides that. If you need more time, that's fine! If you need to drop and let someone else pick this up, that's fine too. Just let me know.

**Comment by PeterBaker0** [comment]
> @PeterBaker0 are you still working on this PR? I see some +1 on my previous comments but no action besides that. If you need more time, that's fine! If you need to drop and let someone else pick this up, that's fine too. Just let me know.

Hey Kaizen - I'd like to do this but I don't see myself having time in the next few weeks. 

Not sure of the process to defer this - I'm happy to close this issue then maybe we can make an issue to track the progress so far and what was missing from this PR, so someone could come and pick it up (or me if I get time)?

**Comment by corymhall** [comment]
> Not sure of the process to defer this - I'm happy to close this PR then maybe we can make an issue to track the progress so far and what was missing from this PR, so someone could come and pick it up (or me if I get time)?

Thanks for letting us know @PeterBaker0! I'll go ahead and close this issue for now. If you ever want to come back to it you can just re-open this PR.

**Comment by johnmarcou** [comment]
@corymhall There is any in progress PR on this subject?

Thank you!

**Comment by PeterBaker0** [comment]
@johnmarcou This was closed as the PR required some more in depth changes which I didn't have time to continue. 

I am unsure if someone has picked up or is working on this. 

You're welcome to continue from where I got to if you wish.

**Pull request closed without merging**
